### ISSUE TITLE
Infinite loop parsing invalid ID response

### ISSUE THREAD

**Issue by jbflanigan** [javaee/javamail#306]
We recently upgraded to 1.6.1 and noticed problems connecting to buggy IMAP servers that use comma between each element in the ID response list. This causes readStringList() to enter an infinite loop and eventually OOM or stack overflow.

While this is technically an incorrect server response, Javamail 1.5.6 was able to successfully parse the response and return the encoded list; i.e. ignoring the stray commas.

Here is a unit test which replicates the problem. If possible, it would be great if Javamail 1.6.x could handle this case.

```java
    @Test
    public void testParseBadResponse() throws IOException, ProtocolException {
        IMAPResponse response = new IMAPResponse("* ID (\"name\", \"test\", \"version\", \"1.0\")");
        String[] list = response.readStringList();
        Assert.assertEquals(4, list.length);
    }
```

**Comment by bshannon** [comment]
It was pure luck that JavaMail didn't complain about that, although clearly looping forever is not good!  :-)

What server produced this response?  Have you reported the bug to the server vendor?

**Comment by jbflanigan** [comment]
It looks like it may come from megamailservers.com, which appears to be a proxy/filtering service. There isn't any good contact info available for the vendor.

If it's easier to throw an exception than try to parse this, that would be just fine for us. We're mostly concerned about the impact on the rest of the JVM when someone happens to access a server with this type of problem.

**Comment by bshannon** [comment]
There's a Contact page on megamailservers.com, and while this isn't abuse it might get you
to someone who can help.  Connecting to their server says:

* OK [CAPABILITY IMAP4REV1 LITERAL+ UTF8=ALL SASL-IR LOGIN-REFERRALS AUTH=CRAM-MD5 AUTH=LOGIN ID] mail18c0 IMAP4rev1 Bigfoot
a id
* ID ("name", "Bigfoot", "version", "1.0", "os", "Linux", "os-version", "2.6", "vendor", "Megamailservers.com", "author", "Derek Snider")

Their web site looks pretty old and has a 2003 copyright so I suspect the software is also
pretty old and may not be maintained anymore.  I can't find a listing for Derek Snider that
seems to be associated with this.

JavaMail 1.5.6 was just ignoring the character after the closing quote, assuming it had to be a
space.  That's why it didn't complain about this.  The obvious fix to prevent it from looping
forever would just result in treating this as an error and throwing an exception.  I'd really rather
not have to add another kludge to ignore a comma after a closing quote.  It would be great if the
owners of this software were responsive and would fix the server, but that seems unlikely.

**Comment by jbflanigan** [comment]
The same OOM can occur with various other invalid responses from other commands and other servers. We'll try to track down and report these to the server author, but would prefer the client not to fail in this way on invalid input.

Here's another example. The unexpected attachment parameter seems to trigger the loop.

```java
        String resp = "* 21 FETCH (ENVELOPE (\"Tue, 13 Feb 2018 14:20:46 +0000\" \"test\" ((NIL NIL \"test\" \"test.com\")) NIL NIL NIL NIL NIL NIL \"<bc96320e-19ba-4f78-a9ca-5442f04ff790@test.example.com>\") BODYSTRUCTURE ((\"text\" \"plain\" (\"charset\" \"UTF-8\") NIL NIL \"quoted-printable\" 169 9 NIL (\"inline\" NIL) NIL NIL) \"mixed\" (\"boundary\" \"--AaZz\") NIL (\"attachment\" (\"filename\" \"test.xml\")) NIL NIL) UID 22966 BODY[HEADER] {1153}\\n";
        IMAPResponse response = new IMAPResponse(resp);
        new FetchResponse(response, new FetchItem[]{});
```

**Comment by jbflanigan** [comment]
any possibility of a fix to at least throw an error instead of loop and eventual OOM? our service connects to many remote servers concurrently; but a single connection to one bad server causes the entire service to hang due to OOM. where practical, we've tried to get in touch with authors, but many servers are poorly maintained.

**Comment by bshannon** [comment]
Yes, I've fixed this for JavaMail 1.6.2.  I'll be pushing out a snapshot release soon, and a final release probably later this summer.

**Comment by jbflanigan** [comment]
thank you!
